{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count alignment is 2000000\n",
      "\n",
      "Number of perfect alignments is 1702765\n",
      "\n",
      "First ten reads aligned to chromosome:\n",
      "1.\t2L\n",
      "2.\t2L\n",
      "3.\t2L\n",
      "4.\t2L\n",
      "5.\t2L\n",
      "6.\t2L\n",
      "7.\t2L\n",
      "8.\t2L\n",
      "9.\t2L\n",
      "10.\t2L\n",
      "Average MAPQ scores across all reads = 59.6407825\n",
      "\n",
      "Total number of reads that start their alignment on chromosome 2L between 10k and 20k base inclusive: 7550\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os, sys\n",
    "\n",
    "with open(\"/Users/cmdb/qbb2020-answers/day1-evening/SRR072893.sam\") as samFile:\n",
    "\tlines = samFile.readlines()\n",
    "\t#count number of alignments\n",
    "\tcount_alignment = 0\n",
    "\t#count number of perfect alignments\n",
    "\tcount_perfect_alignment = 0\n",
    "\t#for the average of mapq values\n",
    "\ttotal_mapq = 0\n",
    "\tmapq_count = 0\n",
    "\t#count number of reads that start alignment on chromosome 2L between 10k and 20k\n",
    "\tchrom_2L_count = 0\n",
    "\tfor line in lines:\n",
    "\t\t#count number of alignments\n",
    "\t\tline = line.strip()\n",
    "\t\tsplit_line = line.split(\"\\t\")\n",
    "\t\t#The alignment result is shown in FLAG format in split_line[1]\n",
    "\t\tif \"4\" in split_line[1]:\n",
    "\t\t\tcount_alignment = count_alignment + 0\n",
    "\t\telse:\n",
    "\t\t\tcount_alignment = count_alignment + 1\n",
    "\n",
    "\t\t#count the number of alignments that match perfectly to the genome\n",
    "\t\tif \"NM:i:0\" in line:\n",
    "\t\t\tcount_perfect_alignment = count_perfect_alignment + 1\n",
    "\n",
    "\t\t#get a list of the first 10 chromosomes\n",
    "\t\tfirst_ten = 0\n",
    "\t\tfirst_ten_chromosome_list = []\n",
    "\t\twhile first_ten < 10:\n",
    "\t\t\tfirst_ten_chromosome_list.append(split_line[2])\n",
    "\t\t\tfirst_ten = first_ten + 1\n",
    "\n",
    "\t\t#calculate the average mapq score across all reads (mapq is found in split_line[4])\n",
    "\t\ttotal_mapq = int(split_line[4]) + total_mapq\n",
    "\t\tmapq_count = mapq_count + 1\n",
    "\n",
    "\t\t#count number of reads that start their alignment on chromosome 2L between base 10000 and 20000\n",
    "\t\tif \"2L\" == split_line[2]:\n",
    "\t\t\tif int(split_line[3]) >= 10000 and int(split_line[3]) <= 20000:\n",
    "\t\t\t\tchrom_2L_count = chrom_2L_count + 1\n",
    "\n",
    "\n",
    "#first question\n",
    "print(\"Count alignment is \" + str(count_alignment) + \"\\n\")\n",
    "\n",
    "#second question\n",
    "print(\"Number of perfect alignments is \" + str(count_perfect_alignment) + \"\\n\")\n",
    "\n",
    "#third question\n",
    "print(\"First ten reads aligned to chromosome:\")\n",
    "for count in range(len(first_ten_chromosome_list)):\n",
    "\tnumber_count = str(count + 1)\n",
    "\tprint(number_count + \".\\t\" + first_ten_chromosome_list[count])\n",
    "\n",
    "#fourth question\n",
    "mapq_average = total_mapq/mapq_count\n",
    "print(\"Average MAPQ scores across all reads = \" + str(mapq_average) + \"\\n\")\n",
    "\n",
    "#fifth question\n",
    "print(\"Total number of reads that start their alignment on chromosome 2L between 10k and 20k base inclusive: \" + str(chrom_2L_count) + \"\\n\")\n",
    "\n",
    "#------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "#Final output:\n",
    "#Count alignment is 2000000\n",
    "\n",
    "#Number of perfect alignments is 1702765\n",
    "\n",
    "#First ten reads aligned to chromosome:\n",
    "#1.\t2L\n",
    "#2.\t2L\n",
    "#3.\t2L\n",
    "#4.\t2L\n",
    "#5.\t2L\n",
    "#6.\t2L\n",
    "#7.\t2L\n",
    "#8.\t2L\n",
    "#9.\t2L\n",
    "#10.\t2L\n",
    "#Average MAPQ scores across all reads = 59.6407825\n",
    "\n",
    "#Total number of reads that start their alignment on chromosome 2L between 10k and 20k base inclusive: 7550\n",
    "#------------------------------------------------------------------------------------------------------------------------------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
